### PR TITLE
Migrate datadog to pythonic resources

### PR DESCRIPTION
**Pull request by schrockn**
## Summary & Motivation

Migrate dagster-datadog to use pythonic resources system

## How I Tested These Changes

Included unit tests

### CODE REVIEW

**Comment by schrockn** [comment]
Current dependencies on/for this PR:
* master
  * **PR #13383** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/13383" target="_blank">Graphite</a> 
    * **PR #13385** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/13385" target="_blank">Graphite</a> 
      * **PR #13387** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/13387" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/13387?utm_source=stack-comment).

**Comment by schrockn** [comment]
I'm particularly interested in bikeshedding on this one in terms of the naming conventions around resources and their factories.

Questions I have

* Should something that inherits from ConfigurableResourceFactory be called a Resource or a Resource Factory
* Should a class created by a ResourceFactory have Resource in its name?


In this PR I settled on the following. I don't have high conviction that this is right.

* `class DatadogClientResource(ConfigurableResourceFactory[DataDogClient]):`
* `class DataDogClient:`

This optimizes for these code snippets:

```
    job_for_datadog_op.execute_in_process(
        resources={"datadog_client": DatadogClientResource(api_key="FOO", app_key="BAR")}
    )
```

```

    @op
    def datadog_op(datadog_client: Resource[DataDogClient]):
 ```
 
 Notes:
 
 * If one names DataDogClient DataDogResource then you have `def datadog_op(datadog_resource: Resource[DataDogResource]):`which isn't great.
 * Similarly the Definitions declaration is a little worse with `resources={"datadog_client": DatadogResourceFactory(api_key="FOO", app_key="BAR")}`

(Please forgive DataDog versus Datadog here will fix in diff) 

 However I think there are some pretty obvious downsides as user is more liable to annotate with DataDogResource directly

**Comment by jamiedemaria** [comment]
Just to make sure i'm understanding the system correctly, if we have a non-factory resource  and I want to use it in an asset, I would do this?

```
class MyResource(ConfigurableResource)
   foo: str

@asset 
def my_asset(my_resource: Resource[MyResource]):
   ...

defs = Definitions(
    assets=[my_asset],
    resources={"my_resource": MyResource(foo="bar")
)
```

but if it were a factory, it would look like this?

```
class MyClient:
    def __init__(self, foo):
        ...
   ...

class MyResource(ConfigurableResourceFactory)
   foo: str

   def create_resource():
       return MyClient(foo=foo)

@asset 
def my_asset(my_resource: Resource[MyClient]):
   ...


defs = Definitions(
    assets=[my_asset],
    resources={"my_resource": MyResource(foo="bar")
)
```

**Comment by benpankow** [comment]
> re: nicks comment ^
> 
> Just to make sure i'm understanding the system correctly, if we have a non-factory resource and I want to use it in an asset, I would do this...

This is accurate except for the first case the `Resource[]` annotation isn't needed - since your resource extends `ConfigurableResource` Dagster is smart enough to know it's a resource:

```python

@asset 
def my_asset(my_resource: MyResource):
   ...
```

**Comment by benpankow** [comment]
> I'm particularly interested in bikeshedding on this one in terms of the naming conventions around resources and their factories.
> 
> Questions I have
> 
> * Should something that inherits from ConfigurableResourceFactory be called a Resource or a Resource Factory
> * Should a class created by a ResourceFactory have Resource in its name?
> 
> In this PR I settled on the following. I don't have high conviction that this is right.
> 
> * `class DatadogClientResource(ConfigurableResourceFactory[DataDogClient]):`
> * `class DataDogClient:`

This is the naming scheme I have been preferring as well. I think it makes clear the role of the classes (a client class has methods to interact with the service & does not contain Dagster-specific code, a `Resource` is specifically a Dagster concept with Dagster config).

Things which inherit from `ConfigurableResourceFactory` for simplicity's sake I have been leaving as `Resource`.

**Comment by schrockn** [comment]
> re: nicks comment ^
> 
> Just to make sure i'm understanding the system correctly, if we have a non-factory resource and I want to use it in an asset, I would do this?
> 
> ```
> class MyResource(ConfigurableResource)
>    foo: str
> 
> @asset 
> def my_asset(my_resource: Resource[MyResource]):
>    ...
> 
> defs = Definitions(
>     assets=[my_asset],
>     resources={"my_resource": MyResource(foo="bar")
> )
> ```

If you inherit from ConfigurableResource, you do not need to use the `Resource` annotation.

```
class MyResource(ConfigurableResource)
    foo: str
 
 @asset 
 def my_asset(my_resource: MyResource):
    ...
 
 defs = Definitions(
     assets=[my_asset],
     resources={"my_resource": MyResource(foo="bar")
 )
 ```

File: python_modules/libraries/dagster-datadog/dagster_datadog/resources.py [hunk 1]
```diff
@@ -30,14 +34,19 @@ def __init__(self, api_key, app_key):
             setattr(self, method, getattr(statsd, method))
 
 
+class DataDogClientResource(ConfigurableResourceFactory[DataDogClient]):
+    api_key: str = Field(description="Datadog API key")
```
Comment (benpankow) [inline python_modules/libraries/dagster-datadog/dagster_datadog/resources.py:38]: nit: add description as docstring (could just use the description from `@resource` below)

could also consider linking to docs about API/app key, we do that inconsistently among other resources (https://docs.datadoghq.com/account_management/api-app-keys/)

File: python_modules/libraries/dagster-datadog/dagster_datadog/resources.py [hunk 2]
```diff
@@ -30,14 +34,19 @@ def __init__(self, api_key, app_key):
...
+    def create_resource(self, _init_context) -> DataDogClient:
+        return DataDogClient(self.api_key, self.app_key)
+
+
 @resource(
-    {
-        "api_key": Field(StringSource, description="Datadog API key"),
-        "app_key": Field(StringSource, description="Datadog application key"),
-    },
+    config_schema=infer_schema_from_config_class(DataDogClientResource),
```
Comment (benpankow) [inline python_modules/libraries/dagster-datadog/dagster_datadog/resources.py:46]: nit: update to new `DataDogClientResource.to_config_schema()`

File: python_modules/libraries/dagster-datadog/dagster_datadog/resources.py [hunk 3]
```diff
@@ -78,6 +87,6 @@ def dd_job():
             )
 
     """
-    return DataDogResource(
+    return DataDogClient(
```
Comment (benpankow) [inline python_modules/libraries/dagster-datadog/dagster_datadog/resources.py:90]: nit: update to new `DataDogClientResource.from_resource_context(context)`

**Review by benpankow — APPROVED** [review]

**Pull request merged**